data/lm/generate_lm.py

```diff
@@ -1,63 +1,147 @@
-               '--arpa', lm_path,
-               '--prune', '0', '0', '1'
+        args.kenlm_bins + 'lmplz',
+        '--order', str(args.arpa_order),
+        '--temp_prefix', args.output_dir,
+        '--memory', args.max_arpa_memory,
+        '--text', data_lower,
+        '--arpa', lm_path,
+        '--prune', '0', '0', '1'
     ])
 
-    vocab_str = '\n'.join(word for word, count in counter.most_common(500000))
-    with open('librispeech-vocab-500k.txt', 'w') as fout:
-      fout.write(vocab_str)
-
     # Filter LM using vocabulary of top 500k words
-    print('Filtering ARPA file...')
-    filtered_path = os.path.join(tmp, 'lm_filtered.arpa')
-    subprocess.run(['filter', 'single', 'model:{}'.format(lm_path), filtered_path], input=vocab_str.encode('utf-8'), check=True)
+    print('\nFiltering ARPA file using vocabulary of top-k words ...')
+    filtered_path = os.path.join(args.output_dir, 'lm_filtered.arpa')
+    subprocess.run([
+        args.kenlm_bins + 'filter',
+        'single',
+        'model:{}'.format(lm_path),
+        filtered_path
+    ], input=vocab_str.encode('utf-8'), check=True)
 
     # Quantize and produce trie binary.
-    print('Building lm.binary...')
+    print('\nBuilding lm.binary ...')
+    binary_path = os.path.join(args.output_dir, 'lm.binary')
     subprocess.check_call([
-      'build_binary', '-a', '255',
-                      '-q', '8',
-                      '-v',
-                      'trie',
-                      filtered_path,
-                      'lm.binary'
+        args.kenlm_bins + 'build_binary',
+        '-a', '255',
+        '-q', '8',
+        '-v',
+        'trie',
+        filtered_path,
+        binary_path
     ])
 
+
+# ======================================================================================================================
+
+def main():
+    parser = argparse.ArgumentParser(
+        description='Generate an lm.binary and top-k vocab for DeepSpeech.'
+    )
+    parser.add_argument(
+        '--input_txt',
+        help='File path to a .txt with sample sentences',
+        type=str,
+        required=True
+    )
+    parser.add_argument(
+        '--output_dir',
+        help='Directory path for the output',
+        type=str,
+        required=True
+    )
+    parser.add_argument(
+        '--top_k',
+        help='Use top_k most frequent words for the vocab.txt file',
+        type=int,
+        default=500000
+    )
+    parser.add_argument(
+        '--kenlm_bins',
+        help='File path to the kenlm binaries lmplz, filter and build_binary',
+        type=str,
+        default='/DeepSpeech/native_client/kenlm/build/bin/'
+    )
+    parser.add_argument(
+        '--arpa_order',
+        help='Order of k-grams in arpa-file generation',
+        type=int,
+        default=5
+    )
+    parser.add_argument(
+        '--max_arpa_memory',
+        help='Maximum allowed memory usage in arpa-file generation',
+        type=str,
+        default='75%'
+    )
```

Comment:
Every default here is arbitrary. Make them optional and document the exact command line used to generate the model in `data/lm/kenlm.scorer`.

Reply:
I tried to use the values you used to generate the model. The path to the kenlm bins is the installation path in the docker container.

Not sure what you mean with optional, if you don't add the parameter to your command, the default values are used, else it will be overwritten with the value from the command.

Reply:
Sorry, I meant to say make them required.

Reply:
Without default values? So that the user has to fill in a value for each single parameter?

Reply:
Correct, and then document the invocation that has the values used for the model in the repo.

Reply:
Ok, i can do this.
Why do you think that is better than using default values? I would prefer default values because i think its more readable, add only flags for params i want to change or which are needed to run this program. Like its also handled in `DeepSpeech.py`.

Reply:
Because other users might have different needs? This is the goal of this PR, right? To make the script more flexible.

Reply:
Yes, but with both of our ways you get a flexible script. I only see a difference in readability and usability. For example in my version i only need to add `--top_k 300000` if i am not happy with `--top_k 500000`. In your version i always have to add  `--top_k 500000`.
Personally i prefer the shorter commands, but its only a design issue. If you think the users should list all parameters, i will move the default values to the example call for the model used here.